src/Http/Http.Extensions/src/RequestDelegateFactory.cs

```diff
@@ -2258,7 +2259,12 @@ private static Task WriteJsonResponse(HttpResponse response, object? value, Json
         // https://github.com/dotnet/aspnetcore/issues/43894
         // https://docs.microsoft.com/en-us/dotnet/standard/serialization/system-text-json-polymorphism
         return HttpResponseJsonExtensions.WriteAsJsonAsync(response, value, value is null ? typeof(object) : value.GetType(), options, default);
+    }
 
+    // Only for use with structs, use WriteJsonResponse for classes to preserve polymorphism
+    private static Task WriteJsonResponseOfT<T>(HttpResponse response, T value, JsonSerializerOptions? options)
```

Comment:
If this is only used for structs, should it have a struct constraint? Probably doesn't make any difference at runtime.
```suggestion
    private static Task WriteJsonResponseOfT<T>(HttpResponse response, T value, JsonSerializerOptions? options) where T : struct
```

Reply:
That breaks when using a nullable struct, the debug assert should be good enough.
https://learn.microsoft.com/dotnet/csharp/programming-guide/generics/constraints-on-type-parameters